In [11]:
import pandas as pd
path = "HeroesOfPymoli/purchase_data.json"
pymoli_df = pd.read_json(path)

pymoli_df.head()

Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN  
0     Aelalis34  
1        Eolo46  
2   Assastnya25  
3  Pheusrical25  
4        Aela59

In [12]:
#Total players
player_count = pymoli_df["SN"].nunique()

player_count_df = pd.DataFrame([{"Total Players" : player_count}])
player_count_df

Total Players
0            573

In [13]:
#question: how to pass multiple columns on the subset(sequence of labels). Should pass SN, age and gender
player_count = pymoli_df.drop_duplicates(subset='SN', keep='first', inplace=False)
unique_player_count = player_count["SN"].count()

player_count_df = pd.DataFrame([{"Total Players" : unique_player_count}])
player_count_df

Total Players
0            573

In [14]:
#**Purchasing Analysis (Total)**

#* Number of Unique Items
#* Average Purchase Price
#* Total Number of Purchases
#* Total Revenue

unique_items = pymoli_df["Item ID"].nunique()

average_price = round(pymoli_df["Price"].mean(),2)

total_purchases = pymoli_df["Price"].count()

revenue = pymoli_df["Price"].sum()

purchasing_analysis_df = pd.DataFrame([{"Number of Unique Items": unique_items, "Average Purchase Price": average_price,
                                       "Total Number of Purchases": total_purchases, "Total Revenue": revenue}])

purchasing_analysis_df['Average Purchase Price']=purchasing_analysis_df['Average Purchase Price'].map('${:,.2f}'.format)
purchasing_analysis_df['Total Revenue']=purchasing_analysis_df['Total Revenue'].map('${:,.2f}'.format)

purchasing_analysis_df


Average Purchase Price  Number of Unique Items  Total Number of Purchases  \
0                  $2.93                     183                        780   

  Total Revenue  
0     $2,286.33

In [ ]:
#**Gender Demographics**

#* Percentage and Count of Male Players
#* Percentage and Count of Female Players
#* Percentage and Count of Other / Non-Disclosed

all_male = pymoli_df['Gender']=='Male'
all_female = pymoli_df['Gender']=='Female'
all_other = (pymoli_df['Gender']!='Male') & (pymoli_df['Gender']!='Female')
gender_index = ['Male', 'Female', 'Other / Non-Disclosed']

male_count =  pymoli_df[all_male]['SN'].nunique()
female_count = pymoli_df[all_female ]['SN'].nunique()
other_count = pymoli_df[all_other]['SN'].nunique()
gender_count = pd.Series([male_count, female_count, other_count], index=gender_index)

male_perc = round(male_count/player_count * 100, 2)
female_perc = round(female_count/player_count * 100, 2)
other_perc = round(other_count/player_count * 100, 2)
gender_perc = pd.Series([male_perc, female_perc, other_perc], index=gender_index)

gender_df = pd.DataFrame({'Total Count':gender_count, 'Percentage of Players':gender_perc})

gender_df['Percentage of Players']=gender_df['Percentage of Players'].map('{:,.2f}%'.format)
gender_df

In [18]:
#**Purchasing Analysis (Gender)** 

#* The below each broken by gender
#  * Purchase Count
#  * Average Purchase Price
#  * Total Purchase Value
#  * Normalized Totals

#take the *Total Purchase Value* for each age group and divide it by the *Total Count* of the respective age group. 
#Essentially the average total per play purchased by group.

purchase_count_male = pymoli_df[all_male]['SN'].nunique()
purchase_count_female = pymoli_df[all_female]['SN'].nunique()
purchase_count_other = pymoli_df[all_other]['SN'].nunique()

gender_purchase = pd.Series([purchase_count_male, purchase_count_female, purchase_count_other],index=gender_index)


average_male_price = pymoli_df[all_male]["Price"].mean()
average_female_price = pymoli_df[all_female]["Price"].mean()
average_other_price = pymoli_df[all_other]["Price"].mean()

average_price_per_gender = pd.Series([average_male_price, average_female_price, average_other_price],index=gender_index)


total_male_price = pymoli_df[all_male]["Price"].sum()
total_female_price = pymoli_df[all_female]["Price"].sum()
total_other_price = pymoli_df[all_other]["Price"].sum()

total_purchase_value_per_gender = pd.Series([total_male_price, total_female_price, total_other_price],index=gender_index)

normalized_male = round(total_male_price/purchase_count_male,2)
normalized_female = round(total_female_price/purchase_count_female,2)
normalized_totals = round(total_other_price/purchase_count_other,2)

normalized_per_gender = pd.Series([normalized_male, normalized_female, normalized_totals], index=gender_index)

purchasing_analysis_df = pd.DataFrame({"Purchase Count":gender_purchase, "Average Purchase Price": average_price_per_gender, 
                                        "Total Purchase Value": total_purchase_value_per_gender, "Normalized Totals":normalized_per_gender})

purchasing_analysis_df['Average Purchase Price']=purchasing_analysis_df['Average Purchase Price'].map('${:,.2f}'.format)
purchasing_analysis_df['Total Purchase Value']=purchasing_analysis_df['Total Purchase Value'].map('${:,.2f}'.format)

purchasing_analysis_df


Average Purchase Price  Normalized Totals  \
Male                                   $2.95               4.02   
Female                                 $2.82               3.83   
Other / Non-Disclosed                  $3.25               4.47   

                       Purchase Count Total Purchase Value  
Male                              465            $1,867.68  
Female                            100              $382.91  
Other / Non-Disclosed               8               $35.74

In [ ]:
pymoli_df["Age"].max()

In [ ]:
pymoli_df["Age"].min()

In [9]:
#**Age Demographics**

#* The below each broken into bins of 4 years (i.e. >10, 10-14, 15-19, etc.) 
 # * Purchase Count
 # * Average Purchase Price
 # * Total Purchase Value
 # * Normalized Totals

bins = [0, 10, 14, 19, 24, 29, 34, 39, 100]
# the outter item is inclusive on the right

# Create labels for these bins
bins_labels = ["< 10", "10 to 14", "15 to 19", "20 to 24", "25 to 29", "30 to 34", "35 to 39", "40+"]

# Slice the data and place it into bins
pd.cut(pymoli_df["Age"], bins, labels=bins_labels)

pymoli_df["Age Demographics"] = pd.cut(pymoli_df["Age"], bins, labels=bins_labels)
pymoli_df.head()



Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN Age Demographics  
0     Aelalis34         35 to 39  
1        Eolo46         20 to 24  
2   Assastnya25         30 to 34  
3  Pheusrical25         20 to 24  
4        Aela59         20 to 24

In [10]:
purchase_by_age = pymoli_df.groupby('Age Demographics')

purchase_count = purchase_by_age['Price'].count()
average_purchase = purchase_by_age['Price'].mean()
total_age_purchase = purchase_by_age['Price'].sum()
#normalized_totals = amount spent per age group / unique number of people per age group nunique


age_demographics_df = pd.DataFrame({'Purchase Count':purchase_count, 'Average Purchase Price':average_purchase,'Total Purchase Value':total_age_purchase})

age_demographics_df['Average Purchase Price']=age_demographics_df['Average Purchase Price'].map('${:,.2f}'.format)
age_demographics_df['Total Purchase Value']=age_demographics_df['Total Purchase Value'].map('${:,.2f}'.format)

age_demographics_df

Average Purchase Price  Purchase Count Total Purchase Value
Age Demographics                                                            
< 10                              $3.02              32               $96.62
10 to 14                          $2.70              31               $83.79
15 to 19                          $2.91             133              $386.42
20 to 24                          $2.91             336              $978.77
25 to 29                          $2.96             125              $370.33
30 to 34                          $3.08              64              $197.25
35 to 39                          $2.84              42              $119.40
40+                               $3.16              17               $53.75

In [7]:
#**Top Spenders**

#* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
 # * SN
  #* Purchase Count
  #* Average Purchase Price
  #* Total Purchase Value


group_spenders = pymoli_df.groupby("SN")
index=group_spenders

spenders_purchase_count = group_spenders['Price'].count()
average_purchase_price_by_spender = group_spenders['Price'].mean()
total_purchase_by_spender = group_spenders['Price'].sum()

top_spenders_df = pd.DataFrame({'Purchase Count':spenders_purchase_count, 'Average Purchase Price':average_purchase_price_by_spender,'Total Purchase Value':total_purchase_by_spender})

sorted_top_spenders = top_spenders_df.sort_values("Total Purchase Value", ascending=False)

sorted_top_spenders['Average Purchase Price']=sorted_top_spenders['Average Purchase Price'].map('${:,.2f}'.format)
sorted_top_spenders['Total Purchase Value']=sorted_top_spenders['Total Purchase Value'].map('${:,.2f}'.format)

sorted_top_spenders.head()


Average Purchase Price  Purchase Count Total Purchase Value
SN                                                                     
Undirrala66                  $3.41               5               $17.06
Saedue76                     $3.39               4               $13.56
Mindimnya67                  $3.18               4               $12.74
Haellysu29                   $4.24               3               $12.73
Eoda93                       $3.86               3               $11.58

In [ ]:
#**Most Popular Items**

#* Identify the 5 most popular items by purchase count, then list (in a table):
#  * Item ID
#  * Item Name
#  * Purchase Count
#  * Item Price
#  * Total Purchase Value

item_count=pd.DataFrame(pymoli_df.groupby(['Item ID','Item Name']) ["Price"].count())
item_mean=pd.DataFrame(pymoli_df.groupby(["Item ID", "Item Name"])["Price"].mean())
item_total=pd.DataFrame(pymoli_df.groupby(["Item ID", "Item Name"])["Price"].sum())

#create dataframe with groupby objects
top_items=pd.DataFrame(
    {"Purchase Count": item_count['Price'], 
     "Item Price": item_mean['Price'],
     "Total Purchase Value": item_total['Price']})

top_items=top_items.sort_values(["Purchase Count"], ascending=False)

top_items['Item Price']=top_items['Item Price'].map('${:,.2f}'.format)
top_items['Total Purchase Value']=top_items['Total Purchase Value'].map('${:,.2f}'.format)

top_items.head()

In [6]:
#**Most Profitable Items**

#* Identify the 5 most profitable items by total purchase value, then list (in a table):
 # * Item ID
  #* Item Name
  #* Purchase Count
  #* Item Price
  #* Total Purchase Value


item_count=pd.DataFrame(pymoli_df.groupby(['Item ID','Item Name']) ["Price"].count())
item_mean=pd.DataFrame(pymoli_df.groupby(["Item ID", "Item Name"])["Price"].mean())
item_total=pd.DataFrame(pymoli_df.groupby(["Item ID", "Item Name"])["Price"].sum())

#create dataframe with groupby objects

top_profit=pd.DataFrame(
    {"Purchase Count": item_count['Price'], 
     "Item Price": item_mean['Price'],
     "Total Purchase Value": item_total['Price']})

top_profit=top_profit.sort_values(["Total Purchase Value"], ascending=False)

top_profit['Item Price']=top_profit['Item Price'].map('${:,.2f}'.format)
top_profit['Total Purchase Value']=top_profit['Total Purchase Value'].map('${:,.2f}'.format)

top_profit.head()

#most_profit=pymoli_df.groupby(['Item ID', 'Item Name', 'Price']).Price.sum().nlargest(5)

Item Price  Purchase Count  \
Item ID Item Name                                               
34      Retribution Axe                 $4.14               9   
115     Spectral Diamond Doomblade      $4.25               7   
32      Orenmir                         $4.95               6   
103     Singed Scalpel                  $4.87               6   
107     Splitter, Foe Of Subtlety       $3.61               8   

                                   Total Purchase Value  
Item ID Item Name                                        
34      Retribution Axe                          $37.26  
115     Spectral Diamond Doomblade               $29.75  
32      Orenmir                                  $29.70  
103     Singed Scalpel                           $29.22  
107     Splitter, Foe Of Subtlety                $28.88

Observable Trends:

1) People between 20 to 24 years-old tend to purchase more often game items
2) The majority of players is male
3) Although non-identified gender showed the highest average purchase, male players generated the highest revenue to this game industry
4) Although the average purchase price of an item is $2.93, the most profitable items are valued between $3.61 and $4.95.